src/liballoc/heap.rs

```diff
@@ -44,6 +45,84 @@ fn check_size_and_alignment(size: usize, align: usize) {
                   align);
 }
 
+#[derive(Copy, Clone, Default, Debug)]
+pub struct HeapAllocator;
+
+unsafe impl Allocator for HeapAllocator {
+    unsafe fn alloc(&mut self, layout: Layout) -> Result<*mut u8, AllocErr> {
+        self.alloc_unchecked(layout.clone())
+            .ok_or(AllocErr::Exhausted { request: layout })
+    }
+
+    unsafe fn dealloc(&mut self, ptr: *mut u8, layout: Layout) {
+        deallocate(ptr, layout.size(), layout.align());
+    }
+
+    unsafe fn usable_size(&self, layout: &Layout) -> (usize, usize) {
+        (layout.size(), usable_size(layout.size(), layout.align()))
+    }
+
+    unsafe fn realloc(&mut self,
+                      ptr: *mut u8,
+                      layout: Layout,
+                      new_layout: Layout)
+                      -> Result<*mut u8, AllocErr>
+    {
+        if layout.align() != new_layout.align() {
+            return Err(AllocErr::Unsupported { details: "mismatched input layouts" });
+        }
+        self.realloc_unchecked(ptr, layout, new_layout.clone())
+            .ok_or(AllocErr::Exhausted { request: new_layout })
+    }
+
+    unsafe fn realloc_in_place(&mut self,
+                               ptr: *mut u8,
+                               layout: Layout,
+                               new_layout: Layout)
+                               -> Result<(), CannotReallocInPlace>
+    {
+        if layout.align() != new_layout.align() || new_layout.size() <= layout.size() {
+            // Layout alignments must match. Hard to detect failure if
+            // new size <= old size, so error then too.
+            return Err(CannotReallocInPlace);
+        }
+        let usable = reallocate_inplace(ptr, layout.size(), new_layout.size(), layout.align());
+        if usable >= new_layout.size() { Ok(()) } else { Err(CannotReallocInPlace) }
+    }
+
+    unsafe fn alloc_unchecked(&mut self, layout: Layout) -> Option<*mut u8> {
```

Comment:
the early RFC drafts used `NonZero`, but there was concern about exposing that as part of the API for this type. See e.g. [aturon's comment here](https://github.com/rust-lang/rfcs/pull/1398#issuecomment-202998086)